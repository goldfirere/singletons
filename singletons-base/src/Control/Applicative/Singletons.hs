{-# LANGUAGE DataKinds #-}
{-# LANGUAGE DefaultSignatures #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wno-orphans #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Control.Applicative.Singletons
-- Copyright   :  (C) 2018 Ryan Scott
-- License     :  BSD-style (see LICENSE)
-- Maintainer  :  Ryan Scott
-- Stability   :  experimental
-- Portability :  non-portable
--
-- Defines the promoted and singled versions of the 'Applicative' type class.
--
----------------------------------------------------------------------------

module Control.Applicative.Singletons (
  PApplicative(..), SApplicative(..),
  PAlternative(..), SAlternative(..),
  Sing, SConst(..), Const, GetConst, sGetConst,
  type (<$>), (%<$>), type (<$), (%<$), type (<**>), (%<**>),
  LiftA, sLiftA, LiftA3, sLiftA3, Optional, sOptional,

  -- * Defunctionalization symbols
  PureSym0, PureSym1,
  type (<*>@#@$), type (<*>@#@$$), type (<*>@#@$$$),
  type (*>@#@$),  type (*>@#@$$),  type (*>@#@$$$),
  type (<*@#@$),  type (<*@#@$$),  type (<*@#@$$$),
  EmptySym0, type (<|>@#@$), type (<|>@#@$$), type (<|>@#@$$$),
  ConstSym0, ConstSym1, GetConstSym0, GetConstSym1,
  type (<$>@#@$),  type (<$>@#@$$),  type (<$>@#@$$$),
  type (<$@#@$),   type (<$@#@$$),   type (<$@#@$$$),
  type (<**>@#@$), type (<**>@#@$$), type (<**>@#@$$$),
  LiftASym0,  LiftASym1,  LiftASym2,
  LiftA2Sym0, LiftA2Sym1, LiftA2Sym2, LiftA2Sym3,
  LiftA3Sym0, LiftA3Sym1, LiftA3Sym2, LiftA3Sym3,
  OptionalSym0, OptionalSym1
  ) where

import Control.Applicative
import Control.Monad.Singletons.Internal
import Data.Functor.Const.Singletons
import Data.Functor.Singletons
import Data.Monoid.Singletons
import Data.Ord (Down(..))
import Data.Ord.Singletons
import Data.Singletons.Base.Instances
import Data.Singletons.TH

$(singletonsOnly [d|
  -- -| One or none.
  optional :: Alternative f => f a -> f (Maybe a)
  optional v = Just <$> v <|> pure Nothing

  instance Monoid a => Applicative ((,) a) where
      pure x = (mempty, x)
      (u, f) <*> (v, x) = (u `mappend` v, f x)
      liftA2 f (u, x) (v, y) = (u `mappend` v, f x y)

  instance Applicative Down where
    pure = Down
    Down f <*> Down x = Down (f x)
  |])
