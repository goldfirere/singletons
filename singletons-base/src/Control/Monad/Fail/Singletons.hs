{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Control.Monad.Fail.Singletons
-- Copyright   :  (C) 2019 Ryan Scott
-- License     :  BSD-style (see LICENSE)
-- Maintainer  :  Ryan Scott
-- Stability   :  experimental
-- Portability :  non-portable
--
-- Defines the promoted and singled versions of the 'MonadFail' type class.
--
----------------------------------------------------------------------------

module Control.Monad.Fail.Singletons (
  PMonadFail(..), SMonadFail(..),

  -- * Defunctionalization symbols
  FailSym0, FailSym1
  ) where

import Control.Monad.Singletons.Internal
import Data.Singletons.Base.Instances
import Data.Singletons.TH

$(singletonsOnly [d|
  -- -| When a value is bound in @do@-notation, the pattern on the left
  -- hand side of @<-@ might not match. In this case, this class
  -- provides a function to recover.
  --
  -- A 'Monad' without a 'MonadFail' instance may only be used in conjunction
  -- with pattern that always match, such as newtypes, tuples, data types with
  -- only a single data constructor, and irrefutable patterns (@~pat@).
  --
  -- Instances of 'MonadFail' should satisfy the following law: @fail s@ should
  -- be a left zero for 'Control.Monad.>>=',
  --
  -- @
  -- fail s >>= f  =  fail s
  -- @
  --
  -- If your 'Monad' is also 'Control.Monad.MonadPlus', a popular definition is
  --
  -- @
  -- fail _ = mzero
  -- @
  class Monad m => MonadFail m where
      fail :: String -> m a

  instance MonadFail Maybe where
      fail _ = Nothing

  instance MonadFail [] where
      fail _ = []
  |])
