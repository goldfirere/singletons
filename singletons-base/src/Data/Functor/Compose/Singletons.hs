{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeAbstractions #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wno-orphans #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Functor.Compose.Singletons
-- Copyright   :  (C) 2021 Ryan Scott
-- License     :  BSD-style (see LICENSE)
-- Maintainer  :  Richard Eisenberg (rae@cs.brynmawr.edu)
-- Stability   :  experimental
-- Portability :  non-portable
--
-- Exports the promoted and singled versions of the 'Compose' data type.
--
-----------------------------------------------------------------------------

module Data.Functor.Compose.Singletons (
  -- * The 'Compose' singleton
  Sing, SCompose(..), GetCompose, sGetCompose,

  -- * Defunctionalization symbols
  ComposeSym0, ComposeSym1,
  GetComposeSym0, GetComposeSym1
  ) where

import Control.Applicative
import Control.Applicative.Singletons
import Data.Eq.Singletons
import Data.Foldable.Singletons
import Data.Functor.Compose
import Data.Functor.Singletons
import Data.Ord.Singletons
import Data.Kind
import Data.Semigroup.Singletons
import Data.Singletons
import Data.Singletons.Base.Instances (SList(..), (:@#@$), NilSym0)
import Data.Singletons.TH
import Data.Traversable.Singletons

{-
In order to keep the type arguments to Compose poly-kinded and with inferred
specificities, we define the singleton version of Compose, as well as its
defunctionalization symbols, by hand. This is very much in the spirit of the
code in Data.Functor.Const.Singletons. (See the comments above SConst in that
module for more details on this choice.)
-}
infixr 9 `SCompose`
type SCompose :: Compose f g a -> Type
data SCompose :: Compose f g a -> Type where
  SCompose :: forall f g a (x :: f (g a)).
              Sing x -> SCompose ('Compose @f @g @a x)
type instance Sing = SCompose
instance SingI x => SingI ('Compose x) where
  sing = SCompose sing
instance SingI1 'Compose where
  liftSing = SCompose

infixr 9 `ComposeSym0`
type ComposeSym0 :: f (g a) ~> Compose f g a
data ComposeSym0 z
type instance Apply ComposeSym0 x = 'Compose x
instance SingI ComposeSym0 where
  sing = singFun1 SCompose

infixr 9 `ComposeSym1`
type ComposeSym1 :: f (g a) -> Compose f g a
type family ComposeSym1 x where
  ComposeSym1 x = 'Compose x

$(singletonsOnly [d|
  getCompose :: Compose f g a -> f (g a)
  getCompose (Compose x) = x

  deriving instance Eq (f (g a)) => Eq (Compose f g a)
  deriving instance Ord (f (g a)) => Ord (Compose f g a)

  instance (Functor f, Functor g) => Functor (Compose f g) where
      fmap f (Compose x) = Compose (fmap (fmap f) x)
      a <$ (Compose x) = Compose (fmap (a <$) x)

  instance (Foldable f, Foldable g) => Foldable (Compose f g) where
      foldMap f (Compose t) = foldMap (foldMap f) t

  instance (Traversable f, Traversable g) => Traversable (Compose f g) where
      traverse f (Compose t) = Compose <$> traverse (traverse f) t

  instance (Applicative f, Applicative g) => Applicative (Compose f g) where
      pure x = Compose (pure (pure x))
      Compose f <*> Compose x = Compose (liftA2 (<*>) f x)
      liftA2 f (Compose x) (Compose y) =
        Compose (liftA2 (liftA2 f) x y)

  instance (Alternative f, Applicative g) => Alternative (Compose f g) where
      empty = Compose empty
      Compose x <|> Compose y = Compose (x <|> y)
  |])
