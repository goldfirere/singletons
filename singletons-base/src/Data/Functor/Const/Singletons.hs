{-# LANGUAGE DataKinds #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE PolyKinds #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE StandaloneKindSignatures #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
{-# OPTIONS_GHC -Wno-orphans #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  Data.Functor.Const.Singletons
-- Copyright   :  (C) 2018 Ryan Scott
-- License     :  BSD-style (see LICENSE)
-- Maintainer  :  Richard Eisenberg (rae@cs.brynmawr.edu)
-- Stability   :  experimental
-- Portability :  non-portable
--
-- Exports the promoted and singled versions of the 'Const' data type.
--
-----------------------------------------------------------------------------

module Data.Functor.Const.Singletons (
  -- * The 'Const' singleton
  Sing, SConst(..), GetConst, sGetConst,

  -- * Defunctionalization symbols
  ConstSym0, ConstSym1,
  GetConstSym0, GetConstSym1
  ) where

import Control.Applicative
import Control.Monad.Singletons.Internal
import Data.Eq.Singletons
import Data.Foldable.Singletons
import Data.Kind (Type)
import Data.Monoid.Singletons
import Data.Ord.Singletons
import Data.Semigroup.Singletons.Internal
import Data.Singletons
import Data.Singletons.Base.Instances hiding (FoldlSym0, sFoldl)
import Data.Singletons.Base.Enum
import Data.Singletons.TH
import GHC.Base.Singletons
  hiding ( Const, ConstSym0, ConstSym1
         , Foldr, FoldrSym0, sFoldr )
import GHC.Num.Singletons
import Text.Show.Singletons

{-
Const's argument `b` is poly-kinded, and as a result, we have a choice as to
what singleton type to give it. We could use either

1. type SConst :: forall {k :: Type} (a :: Type) (b :: k).    Const a b -> Type
2. type SConst :: forall             (a :: Type) (b :: Type). Const a b -> Type

Option (1) is the more permissive one, so we opt for that. However, singletons-th's
TH machinery does not jive with this option, since the SingKind instance it
tries to generate:

  instance (SingKind a, SingKind b) => SingKind (Const a b) where
    type Demote (Const a b) = Const (Demote a) (Demote b)

Assumes that `b` is of kind Type. Until we get a more reliable story for
poly-kinded Sing instances (see #150), we simply write the singleton type by
hand.
-}
type SConst :: Const a b -> Type
data SConst :: Const a b -> Type where
  SConst :: Sing a -> SConst ('Const a)
type instance Sing = SConst
instance SingKind a => SingKind (Const a b) where
  type Demote (Const a b) = Const (Demote a) b
  fromSing (SConst sa) = Const (fromSing sa)
  toSing (Const a) = withSomeSing a $ SomeSing . SConst
instance SingI a => SingI ('Const a) where
  sing = SConst sing

$(genDefunSymbols [''Const])
instance SingI ConstSym0 where
  sing = singFun1 SConst

$(singletonsOnly [d|
  getConst :: Const a b -> a
  getConst (Const x) = x

  deriving instance Bounded a => Bounded (Const a b)
  deriving instance Eq      a => Eq      (Const a b)
  deriving instance Ord     a => Ord     (Const a b)

  -- deriving instance Enum a => Enum (Const a b)
  instance Enum a => Enum (Const a b) where
    succ (Const x)     = Const (succ x)
    pred (Const x)     = Const (pred x)
    toEnum i           = Const (toEnum i)
    fromEnum (Const x) = fromEnum x
    enumFromTo (Const x) (Const y) = map Const (enumFromTo x y)
    enumFromThenTo (Const x) (Const y) (Const z) =
        map Const (enumFromThenTo x y z)

  -- deriving instance Monoid a => Monoid (Const a b)
  instance Monoid a => Monoid (Const a b) where
    mempty = Const mempty

  -- deriving instance Num a => Num (Const a b)
  instance Num a => Num (Const a b) where
    Const x + Const y = Const (x + y)
    Const x - Const y = Const (x - y)
    Const x * Const y = Const (x * y)
    negate (Const x)  = Const (negate x)
    abs    (Const x)  = Const (abs    x)
    signum (Const x)  = Const (signum x)
    fromInteger n     = Const (fromInteger n)

  -- deriving instance Semigroup a => Semigroup (Const a b)
  instance Semigroup a => Semigroup (Const a b) where
    Const x <> Const y = Const (x <> y)

  -- -| This instance would be equivalent to the derived instances of the
  -- 'Const' newtype if the 'runConst' field were removed
  instance Show a => Show (Const a b) where
      showsPrec d (Const x) = showParen (d > 10) $
                              showString "Const " . showsPrec 11 x

  deriving instance Functor (Const m)
  deriving instance Foldable (Const m)

  instance Monoid m => Applicative (Const m) where
      pure _ = Const mempty
      liftA2 _ (Const x) (Const y) = Const (x `mappend` y)
      Const x <*> Const y = Const (x `mappend` y)
  |])
