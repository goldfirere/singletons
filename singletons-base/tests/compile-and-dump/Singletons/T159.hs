module T159 where

import Data.Singletons.TH

data T0 = A | B | C | D | E | F
  deriving (Show)

data T1 = N1 | C1 T0 T1 | T0 :&& T1
  deriving (Show)

infixr 5 `C1`
infixr 5 :&&

genSingletons [''T0, ''T1]

singletons [d|
  data T2 = N2 | C2 T0 T2 | T0 :|| T2

  infixr 5 `C2`
  infixr 5 :||
  |]

t1 :: T1
t1 = fromSing $ SA `SC1` SB `SC1` SD :%&& SE :%&& SF `SC1` SN1

t2 :: T2
t2 = fromSing $ SA `SC2` SB `SC2` SD :%|| SE :%|| SF `SC2` SN2
