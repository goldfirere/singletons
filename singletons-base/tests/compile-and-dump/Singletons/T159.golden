Singletons/T159.hs:0:0:: Splicing declarations
    genSingletons [''T0, ''T1]
  ======>
    type ASym0 :: T0
    type family ASym0 :: T0 where
      ASym0 = 'A
    type BSym0 :: T0
    type family BSym0 :: T0 where
      BSym0 = 'B
    type CSym0 :: T0
    type family CSym0 :: T0 where
      CSym0 = 'C
    type DSym0 :: T0
    type family DSym0 :: T0 where
      DSym0 = 'D
    type ESym0 :: T0
    type family ESym0 :: T0 where
      ESym0 = 'E
    type FSym0 :: T0
    type family FSym0 :: T0 where
      FSym0 = 'F
    type ST0 :: T0 -> Type
    data ST0 :: T0 -> Type
      where
        SA :: ST0 ('A :: T0)
        SB :: ST0 ('B :: T0)
        SC :: ST0 ('C :: T0)
        SD :: ST0 ('D :: T0)
        SE :: ST0 ('E :: T0)
        SF :: ST0 ('F :: T0)
    type instance Sing @T0 = ST0
    type instance Demote T0 = T0
    type instance Promote T0 = T0
    instance SingKind T0 where
      fromSing SA = A
      fromSing SB = B
      fromSing SC = C
      fromSing SD = D
      fromSing SE = E
      fromSing SF = F
      toSing A = SomeSing SA
      toSing B = SomeSing SB
      toSing C = SomeSing SC
      toSing D = SomeSing SD
      toSing E = SomeSing SE
      toSing F = SomeSing SF
    type instance DemoteX @T0 'A = 'A
    type instance DemoteX @T0 'B = 'B
    type instance DemoteX @T0 'C = 'C
    type instance DemoteX @T0 'D = 'D
    type instance DemoteX @T0 'E = 'E
    type instance DemoteX @T0 'F = 'F
    type instance PromoteX @T0 'A = 'A
    type instance PromoteX @T0 'B = 'B
    type instance PromoteX @T0 'C = 'C
    type instance PromoteX @T0 'D = 'D
    type instance PromoteX @T0 'E = 'E
    type instance PromoteX @T0 'F = 'F
    type instance SingKindC @T0 'A = ()
    type instance SingKindC @T0 'B = ()
    type instance SingKindC @T0 'C = ()
    type instance SingKindC @T0 'D = ()
    type instance SingKindC @T0 'E = ()
    type instance SingKindC @T0 'F = ()
    instance SingI 'A where
      sing = SA
    instance SingI 'B where
      sing = SB
    instance SingI 'C where
      sing = SC
    instance SingI 'D where
      sing = SD
    instance SingI 'E where
      sing = SE
    instance SingI 'F where
      sing = SF
    type N1Sym0 :: T1
    type family N1Sym0 :: T1 where
      N1Sym0 = 'N1
    type C1Sym0 :: (~>) T0 ((~>) T1 T1)
    data C1Sym0 :: (~>) T0 ((~>) T1 T1)
      where
        C1Sym0KindInference :: SameKind (Apply C1Sym0 arg) (C1Sym1 arg) =>
                               C1Sym0 a0123456789876543210
    type instance Apply @T0 @((~>) T1 T1) C1Sym0 a0123456789876543210 = C1Sym1 a0123456789876543210
    instance SuppressUnusedWarnings C1Sym0 where
      suppressUnusedWarnings = snd ((,) C1Sym0KindInference ())
    infixr 5 type `C1Sym0`
    type C1Sym1 :: T0 -> (~>) T1 T1
    data C1Sym1 (a0123456789876543210 :: T0) :: (~>) T1 T1
      where
        C1Sym1KindInference :: SameKind (Apply (C1Sym1 a0123456789876543210) arg) (C1Sym2 a0123456789876543210 arg) =>
                               C1Sym1 a0123456789876543210 a0123456789876543210
    type instance Apply @T1 @T1 (C1Sym1 a0123456789876543210) a0123456789876543210 = 'C1 a0123456789876543210 a0123456789876543210
    instance SuppressUnusedWarnings (C1Sym1 a0123456789876543210) where
      suppressUnusedWarnings = snd ((,) C1Sym1KindInference ())
    infixr 5 type `C1Sym1`
    type C1Sym2 :: T0 -> T1 -> T1
    type family C1Sym2 (a0123456789876543210 :: T0) (a0123456789876543210 :: T1) :: T1 where
      C1Sym2 a0123456789876543210 a0123456789876543210 = 'C1 a0123456789876543210 a0123456789876543210
    infixr 5 type `C1Sym2`
    type (:&&@#@$) :: (~>) T0 ((~>) T1 T1)
    data (:&&@#@$) :: (~>) T0 ((~>) T1 T1)
      where
        (::&&@#@$###) :: SameKind (Apply (:&&@#@$) arg) ((:&&@#@$$) arg) =>
                         (:&&@#@$) a0123456789876543210
    type instance Apply @T0 @((~>) T1 T1) (:&&@#@$) a0123456789876543210 = (:&&@#@$$) a0123456789876543210
    instance SuppressUnusedWarnings (:&&@#@$) where
      suppressUnusedWarnings = snd ((,) (::&&@#@$###) ())
    infixr 5 type :&&@#@$
    type (:&&@#@$$) :: T0 -> (~>) T1 T1
    data (:&&@#@$$) (a0123456789876543210 :: T0) :: (~>) T1 T1
      where
        (::&&@#@$$###) :: SameKind (Apply ((:&&@#@$$) a0123456789876543210) arg) ((:&&@#@$$$) a0123456789876543210 arg) =>
                          (:&&@#@$$) a0123456789876543210 a0123456789876543210
    type instance Apply @T1 @T1 ((:&&@#@$$) a0123456789876543210) a0123456789876543210 = '(:&&) a0123456789876543210 a0123456789876543210
    instance SuppressUnusedWarnings ((:&&@#@$$) a0123456789876543210) where
      suppressUnusedWarnings = snd ((,) (::&&@#@$$###) ())
    infixr 5 type :&&@#@$$
    type (:&&@#@$$$) :: T0 -> T1 -> T1
    type family (:&&@#@$$$) (a0123456789876543210 :: T0) (a0123456789876543210 :: T1) :: T1 where
      (:&&@#@$$$) a0123456789876543210 a0123456789876543210 = '(:&&) a0123456789876543210 a0123456789876543210
    infixr 5 type :&&@#@$$$
    type ST1 :: T1 -> Type
    data ST1 :: T1 -> Type
      where
        SN1 :: ST1 ('N1 :: T1)
        SC1 :: forall (n :: T0) (n :: T1).
               (Sing n) -> (Sing n) -> ST1 ('C1 n n :: T1)
        (:%&&) :: forall (n :: T0) (n :: T1).
                  (Sing n) -> (Sing n) -> ST1 ('(:&&) n n :: T1)
    type instance Sing @T1 = ST1
    type instance Demote T1 = T1
    type instance Promote T1 = T1
    instance SingKind T1 where
      fromSing SN1 = N1
      fromSing (SC1 b b) = C1 (fromSing b) (fromSing b)
      fromSing ((:%&&) b b) = (:&&) (fromSing b) (fromSing b)
      toSing N1 = SomeSing SN1
      toSing (C1 (b :: Demote T0) (b :: Demote T1))
        = (\cases (SomeSing c) (SomeSing c) -> SomeSing (SC1 c c))
            (toSing b :: SomeSing T0) (toSing b :: SomeSing T1)
      toSing ((:&&) (b :: Demote T0) (b :: Demote T1))
        = (\cases (SomeSing c) (SomeSing c) -> SomeSing ((:%&&) c c))
            (toSing b :: SomeSing T0) (toSing b :: SomeSing T1)
    type instance DemoteX @T1 'N1 = 'N1
    type instance DemoteX @T1 ('C1 x x) = 'C1 (DemoteX x) (DemoteX x)
    type instance DemoteX @T1 ('(:&&) x x) = '(:&&) (DemoteX x) (DemoteX x)
    type instance PromoteX @T1 'N1 = 'N1
    type instance PromoteX @T1 ('C1 x x) = 'C1 (PromoteX x) (PromoteX x)
    type instance PromoteX @T1 ('(:&&) x x) = '(:&&) (PromoteX x) (PromoteX x)
    type instance SingKindC @T1 'N1 = ()
    type instance SingKindC @T1 ('C1 x x) = (SingKindC x, SingKindC x)
    type instance SingKindC @T1 ('(:&&) x x) = (SingKindC x,
                                                SingKindC x)
    infixr 5 data `SC1`
    infixr 5 data :%&&
    instance SingI 'N1 where
      sing = SN1
    instance (SingI n, SingI n) =>
             SingI ('C1 (n :: T0) (n :: T1)) where
      sing = SC1 sing sing
    instance SingI n => SingI1 ('C1 (n :: T0)) where
      liftSing = SC1 sing
    instance SingI2 'C1 where
      liftSing2 = SC1
    instance SingI (C1Sym0 :: (~>) T0 ((~>) T1 T1)) where
      sing = singFun2 @C1Sym0 SC1
    instance SingI d => SingI (C1Sym1 (d :: T0) :: (~>) T1 T1) where
      sing = singFun1 @(C1Sym1 (d :: T0)) (SC1 (sing @d))
    instance SingI1 (C1Sym1 :: T0 -> (~>) T1 T1) where
      liftSing (s :: Sing (d :: T0))
        = singFun1 @(C1Sym1 (d :: T0)) (SC1 s)
    instance (SingI n, SingI n) =>
             SingI ('(:&&) (n :: T0) (n :: T1)) where
      sing = (:%&&) sing sing
    instance SingI n => SingI1 ('(:&&) (n :: T0)) where
      liftSing = (:%&&) sing
    instance SingI2 '(:&&) where
      liftSing2 = (:%&&)
    instance SingI ((:&&@#@$) :: (~>) T0 ((~>) T1 T1)) where
      sing = singFun2 @(:&&@#@$) (:%&&)
    instance SingI d =>
             SingI ((:&&@#@$$) (d :: T0) :: (~>) T1 T1) where
      sing = singFun1 @((:&&@#@$$) (d :: T0)) ((:%&&) (sing @d))
    instance SingI1 ((:&&@#@$$) :: T0 -> (~>) T1 T1) where
      liftSing (s :: Sing (d :: T0))
        = singFun1 @((:&&@#@$$) (d :: T0)) ((:%&&) s)
Singletons/T159.hs:(0,0)-(0,0): Splicing declarations
    singletons
      [d| infixr 5 :||
          infixr 5 `C2`
          
          data T2 = N2 | C2 T0 T2 | T0 :|| T2 |]
  ======>
    data T2 = N2 | C2 T0 T2 | T0 :|| T2
    infixr 5 `C2`
    infixr 5 :||
    type N2Sym0 :: T2
    type family N2Sym0 :: T2 where
      N2Sym0 = N2
    type C2Sym0 :: (~>) T0 ((~>) T2 T2)
    data C2Sym0 :: (~>) T0 ((~>) T2 T2)
      where
        C2Sym0KindInference :: SameKind (Apply C2Sym0 arg) (C2Sym1 arg) =>
                               C2Sym0 a0123456789876543210
    type instance Apply @T0 @((~>) T2 T2) C2Sym0 a0123456789876543210 = C2Sym1 a0123456789876543210
    instance SuppressUnusedWarnings C2Sym0 where
      suppressUnusedWarnings = snd ((,) C2Sym0KindInference ())
    infixr 5 type `C2Sym0`
    type C2Sym1 :: T0 -> (~>) T2 T2
    data C2Sym1 (a0123456789876543210 :: T0) :: (~>) T2 T2
      where
        C2Sym1KindInference :: SameKind (Apply (C2Sym1 a0123456789876543210) arg) (C2Sym2 a0123456789876543210 arg) =>
                               C2Sym1 a0123456789876543210 a0123456789876543210
    type instance Apply @T2 @T2 (C2Sym1 a0123456789876543210) a0123456789876543210 = C2 a0123456789876543210 a0123456789876543210
    instance SuppressUnusedWarnings (C2Sym1 a0123456789876543210) where
      suppressUnusedWarnings = snd ((,) C2Sym1KindInference ())
    infixr 5 type `C2Sym1`
    type C2Sym2 :: T0 -> T2 -> T2
    type family C2Sym2 (a0123456789876543210 :: T0) (a0123456789876543210 :: T2) :: T2 where
      C2Sym2 a0123456789876543210 a0123456789876543210 = C2 a0123456789876543210 a0123456789876543210
    infixr 5 type `C2Sym2`
    type (:||@#@$) :: (~>) T0 ((~>) T2 T2)
    data (:||@#@$) :: (~>) T0 ((~>) T2 T2)
      where
        (::||@#@$###) :: SameKind (Apply (:||@#@$) arg) ((:||@#@$$) arg) =>
                         (:||@#@$) a0123456789876543210
    type instance Apply @T0 @((~>) T2 T2) (:||@#@$) a0123456789876543210 = (:||@#@$$) a0123456789876543210
    instance SuppressUnusedWarnings (:||@#@$) where
      suppressUnusedWarnings = snd ((,) (::||@#@$###) ())
    infixr 5 type :||@#@$
    type (:||@#@$$) :: T0 -> (~>) T2 T2
    data (:||@#@$$) (a0123456789876543210 :: T0) :: (~>) T2 T2
      where
        (::||@#@$$###) :: SameKind (Apply ((:||@#@$$) a0123456789876543210) arg) ((:||@#@$$$) a0123456789876543210 arg) =>
                          (:||@#@$$) a0123456789876543210 a0123456789876543210
    type instance Apply @T2 @T2 ((:||@#@$$) a0123456789876543210) a0123456789876543210 = (:||) a0123456789876543210 a0123456789876543210
    instance SuppressUnusedWarnings ((:||@#@$$) a0123456789876543210) where
      suppressUnusedWarnings = snd ((,) (::||@#@$$###) ())
    infixr 5 type :||@#@$$
    type (:||@#@$$$) :: T0 -> T2 -> T2
    type family (:||@#@$$$) (a0123456789876543210 :: T0) (a0123456789876543210 :: T2) :: T2 where
      (:||@#@$$$) a0123456789876543210 a0123456789876543210 = (:||) a0123456789876543210 a0123456789876543210
    infixr 5 type :||@#@$$$
    infixr 5 data :%||
    infixr 5 data `SC2`
    data ST2 :: T2 -> Type
      where
        SN2 :: ST2 (N2 :: T2)
        SC2 :: forall (n :: T0) (n :: T2).
               (Sing n) -> (Sing n) -> ST2 (C2 n n :: T2)
        (:%||) :: forall (n :: T0) (n :: T2).
                  (Sing n) -> (Sing n) -> ST2 ((:||) n n :: T2)
    type instance Sing @T2 = ST2
    type instance Demote T2 = T2
    type instance Promote T2 = T2
    instance SingKind T2 where
      fromSing SN2 = N2
      fromSing (SC2 b b) = C2 (fromSing b) (fromSing b)
      fromSing ((:%||) b b) = (:||) (fromSing b) (fromSing b)
      toSing N2 = SomeSing SN2
      toSing (C2 (b :: Demote T0) (b :: Demote T2))
        = (\cases (SomeSing c) (SomeSing c) -> SomeSing (SC2 c c))
            (toSing b :: SomeSing T0) (toSing b :: SomeSing T2)
      toSing ((:||) (b :: Demote T0) (b :: Demote T2))
        = (\cases (SomeSing c) (SomeSing c) -> SomeSing ((:%||) c c))
            (toSing b :: SomeSing T0) (toSing b :: SomeSing T2)
    type instance DemoteX @T2 N2 = N2
    type instance DemoteX @T2 (C2 x x) = C2 (DemoteX x) (DemoteX x)
    type instance DemoteX @T2 ((:||) x x) = (:||) (DemoteX x) (DemoteX x)
    type instance PromoteX @T2 N2 = N2
    type instance PromoteX @T2 (C2 x x) = C2 (PromoteX x) (PromoteX x)
    type instance PromoteX @T2 ((:||) x x) = (:||) (PromoteX x) (PromoteX x)
    type instance SingKindC @T2 N2 = ()
    type instance SingKindC @T2 (C2 x x) = (SingKindC x, SingKindC x)
    type instance SingKindC @T2 ((:||) x x) = (SingKindC x,
                                               SingKindC x)
    instance SingI N2 where
      sing = SN2
    instance (SingI n, SingI n) => SingI (C2 (n :: T0) (n :: T2)) where
      sing = SC2 sing sing
    instance SingI n => SingI1 (C2 (n :: T0)) where
      liftSing = SC2 sing
    instance SingI2 C2 where
      liftSing2 = SC2
    instance SingI (C2Sym0 :: (~>) T0 ((~>) T2 T2)) where
      sing = singFun2 @C2Sym0 SC2
    instance SingI d => SingI (C2Sym1 (d :: T0) :: (~>) T2 T2) where
      sing = singFun1 @(C2Sym1 (d :: T0)) (SC2 (sing @d))
    instance SingI1 (C2Sym1 :: T0 -> (~>) T2 T2) where
      liftSing (s :: Sing (d :: T0))
        = singFun1 @(C2Sym1 (d :: T0)) (SC2 s)
    instance (SingI n, SingI n) =>
             SingI ((:||) (n :: T0) (n :: T2)) where
      sing = (:%||) sing sing
    instance SingI n => SingI1 ((:||) (n :: T0)) where
      liftSing = (:%||) sing
    instance SingI2 (:||) where
      liftSing2 = (:%||)
    instance SingI ((:||@#@$) :: (~>) T0 ((~>) T2 T2)) where
      sing = singFun2 @(:||@#@$) (:%||)
    instance SingI d =>
             SingI ((:||@#@$$) (d :: T0) :: (~>) T2 T2) where
      sing = singFun1 @((:||@#@$$) (d :: T0)) ((:%||) (sing @d))
    instance SingI1 ((:||@#@$$) :: T0 -> (~>) T2 T2) where
      liftSing (s :: Sing (d :: T0))
        = singFun1 @((:||@#@$$) (d :: T0)) ((:%||) s)
